sw/device/tests/dif/dif_uart_test.cc

```diff
@@ -0,0 +1,644 @@
+  dif_uart_config_.parity_enable = kDifUartEnable;
+  dif_uart_config_.parity = kDifUartParityOdd;
+
+  EXPECT_WRITE32(UART_CTRL_REG_OFFSET, {
+                                           {UART_CTRL_TX, true},
+                                           {UART_CTRL_RX, true},
+                                           {UART_CTRL_PARITY_EN, true},
+                                           {UART_CTRL_PARITY_ODD, true},
+                                           {UART_CTRL_NCO_OFFSET, 1},
+                                       });
+
+  EXPECT_WRITE32(UART_FIFO_CTRL_REG_OFFSET,
+                 {
+                     {UART_FIFO_CTRL_RXRST, true}, {UART_FIFO_CTRL_TXRST, true},
+                 });
+
+  EXPECT_WRITE32(UART_INTR_ENABLE_REG_OFFSET, 0);
+
+  EXPECT_TRUE(dif_uart_configure(&dif_uart_, &dif_uart_config_));
+}
+
+class WatermarkRxSetTest : public UartTest {};
+
+TEST_F(WatermarkRxSetTest, UartNull) {
+  EXPECT_FALSE(dif_uart_watermark_rx_set(nullptr, kDifUartWatermarkByte1));
+}
+
+/**
+ * Tests the first and the last RX watermark variants.
+ */
+TEST_F(WatermarkRxSetTest, Success) {
+  EXPECT_MASK32(UART_FIFO_CTRL_REG_OFFSET,
+                {
+                    {UART_FIFO_CTRL_RXILVL_OFFSET, UART_FIFO_CTRL_RXILVL_MASK,
+                     UART_FIFO_CTRL_RXILVL_RXLVL1},
+                });
+
+  EXPECT_TRUE(dif_uart_watermark_rx_set(&dif_uart_, kDifUartWatermarkByte1));
+
+  EXPECT_MASK32(UART_FIFO_CTRL_REG_OFFSET,
+                {
+                    {UART_FIFO_CTRL_RXILVL_OFFSET, UART_FIFO_CTRL_RXILVL_MASK,
+                     UART_FIFO_CTRL_RXILVL_RXLVL30},
+                });
+
+  EXPECT_TRUE(dif_uart_watermark_rx_set(&dif_uart_, kDifUartWatermarkByte30));
+}
+
+class WatermarkTxSetTest : public UartTest {};
+
+TEST_F(WatermarkTxSetTest, UartNull) {
+  EXPECT_FALSE(dif_uart_watermark_tx_set(nullptr, kDifUartWatermarkByte1));
+}
+
+/**
+ * Tests the first and the last TX watermark variants.
+ */
+TEST_F(WatermarkTxSetTest, Success) {
+  EXPECT_MASK32(UART_FIFO_CTRL_REG_OFFSET,
+                {
+                    {UART_FIFO_CTRL_TXILVL_OFFSET, UART_FIFO_CTRL_TXILVL_MASK,
+                     UART_FIFO_CTRL_TXILVL_TXLVL1},
+                });
+
+  EXPECT_TRUE(dif_uart_watermark_tx_set(&dif_uart_, kDifUartWatermarkByte1));
+
+  EXPECT_MASK32(UART_FIFO_CTRL_REG_OFFSET,
+                {
+                    {UART_FIFO_CTRL_TXILVL_OFFSET, UART_FIFO_CTRL_TXILVL_MASK,
+                     UART_FIFO_CTRL_TXILVL_TXLVL16},
+                });
+
+  EXPECT_TRUE(dif_uart_watermark_tx_set(&dif_uart_, kDifUartWatermarkByte16));
+}
+
+class BytesSendTest : public UartTest {
+ protected:
+  /**
+   * Sets TX bytes expectations.
+   *
+   * Every sent byte by the "send bytes" routine is expected to result in the
+   * STATUS read of 0 (FIFO not full), and write to WDATA.
+   */
+  void ExpectSendBytes(int num_elements) {
+    for (int i = 0; i < num_elements; ++i) {
+      uint32_t value = static_cast<uint32_t>(kBytesArray[i]);
```

Comment:
Maybe add a check to make sure `i` doesn't go past the end of `kBytesArray`?

Reply:
Good point, will do.

Reply:
Done.